### DISCUSSION TITLE
Why can't I run Avalonia program in Linux DRM with Framework-dependent?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#16519] [Q&A] [answered]
I install .net SDK by this tutorial:
https://learn.microsoft.com/en-us/dotnet/iot/deployment

And it installed successfully:
![image](https://github.com/user-attachments/assets/368ffb6d-2619-47d9-8a48-e93b335a3962)

However, when I run the program. It reports this error:
![image](https://github.com/user-attachments/assets/f93dcc0e-38af-4462-98e9-ff0542262b09)

As the image above, I have installed .net runtime already. Why did it say .net runtime has not been installed? What's wrong with it?

BTW, the program will run well if I publish it with self-contained.  However, I find this cumbersome and would like to use a Framework-dependent solution.

**Comment by thevortexcloud** [comment]
Double check you have compiled for the right architecture/installed the right version of .NET (EG 64bit ARM instead of 32bit). Also make sure you installed the right version of .NET. Some distros also disable certain architectures by default for their package managers (not sure what the Pi OS does?) so you may need to enable them as well by running the dpkg command. 

`dotnet --info` should show you most of what you need.

Failing that see if you can execute the main dll directly with `dotnet AvaloniaApplication3.dll`

**Reply by kyuranger** [reply]
Here is my dotnet --info:
![image](https://github.com/user-attachments/assets/34065f99-5c16-46f6-88a0-2a5a632f529e)

**Reply by kyuranger** [reply]
What's more, there is not AvaloniaApplication.dll in the directory also:
![image](https://github.com/user-attachments/assets/04902115-1621-492f-90e9-edf8f564162c)

Because of this, it cannot be run:
![image](https://github.com/user-attachments/assets/69046538-77a1-40c3-abbd-b57a7eec84b1)

**Reply by LUJIAN2020** [reply]
run ./AvaloniaApplication3

**Reply by kyuranger** [reply]
> run ./AvaloniaApplication3

It doesn't work also.

![image](https://github.com/user-attachments/assets/0867d6b4-cd6a-4680-b909-6cef0bafa353)

**Reply by thevortexcloud** [reply]
What does `echo $DOTNET_ROOT` show?

**Reply by kyuranger** [reply]
> What does `echo $DOTNET_ROOT` show?

Here is the result:
![image](https://github.com/user-attachments/assets/0f4746d6-2be2-4a24-be2a-931382001510)

**Reply by thevortexcloud** [reply]
That's fairly unusual. If you are using the official dotnet installer/packages it typically gets installed into a system directory, not your home directory. But there is also theoretically nothing wrong with that.

Have you rebooted your Pi since you installed .NET? It's possible some environment variables have not been updated properly. They typically get loaded when the shell first loads.

**Reply by kyuranger** [reply]
> That's fairly unusual. If you are using the official dotnet installer/packages it typically gets installed into a system directory, not your home directory. But there is also theoretically nothing wrong with that.
> 
> Have you rebooted your Pi since you installed .NET? It's possible some environment variables have not been updated properly. They typically get loaded when the shell first loads.

I installed it according to Microsoft's official tutorial. Logically, there should be no problem, otherwise dotnet --info would not output the content normally. After installation, I also tried to restart, but the problem still exists. I have asked this question on Stackoverflow now, and I can only wait and see if Microsoft's official staff will answer it.

**Comment by kyuranger** [comment, marked as answer]
Someone on stackoverflow suggested that I change the startup command from
`sudo ./AvaloniaApplication3/AvaloniaApplication3 --drm`
to:
`./AvaloniaApplication3/AvaloniaApplication3 --drm`
Yes, just remove sudo and the program will run normally.
In my understanding, sudo is a command used to execute a program with administrator privileges. I don't quite understand why it affects its operation.

**Reply by kekekeks** [reply]
`sudo` will use root environment, which means that your PATH won't include dotnet "installed" into your user home directory and DOTNET_HOME will be missing.

Use system-wide installation from proper packages.

**Reply by kyuranger** [reply]
I see. Thank you very much.

**Reply by thevortexcloud** [reply]
>  sudo is a command used to execute a program with administrator privileges.

Yes but also no. Sudo actually runs a command _as_ the `root` user by default. You can see that if you run `sudo whoami`. Meaning you are not just running things with a higher privilege level, but also as another user entirely. It just happens that `root` is the account with the highest level of permissions on most Linux systems. Sudo itself however can be used to run things as any user. The closest command that exists on Windows is the `runas` command. Although Microsoft is working on proper sudo support.

https://learn.microsoft.com/en-us/windows/sudo/

On a related note: You can open a proper terminal session as root by doing `sudo su`. The convention is to show a `$` for user sessions and `#` for root sessions.

**Reply by kyuranger** [reply]
> > sudo is a command used to execute a program with administrator privileges.
> 
> Yes but also no. Sudo actually runs a command _as_ the `root` user by default. You can see that if you run `sudo whoami`. Meaning you are not just running things with a higher privilege level, but also as another user entirely. It just happens that `root` is the account with the highest level of permissions on most Linux systems. Sudo itself however can be used to run things as any user. The closest command that exists on Windows is the `runas` command. Although Microsoft is working on proper sudo support.
> 
> https://learn.microsoft.com/en-us/windows/sudo/
> 
> On a related note: You can open a proper terminal session as root by doing `sudo su`. The convention is to show a `$` for user sessions and `#` for root sessions.

Thanks a lot.
